enumerable.rb

```diff
@@ -0,0 +1,170 @@
+    # and return back an Array/Hash depending what it got
+
+    return_element = self.class.new
+
+    my_each do |n|
+      # if it is a Hash then I need to add the key and value if the pair is "selected" (the statement/yield is true)
+      if is_a? Hash
+        return_element[n.to_a[0]] = n.to_a[1] if yield n.to_a[0], n.to_a[1]
+      elsif yield n
+        # If not a Hash and the statement is true push the element to the array
+        # Convert if nested into and elsif.
+        # It is made with an elsif because if not a Hash then an Array, and the only else statement is
+        # if yield is true or false.
+        return_element.push(n)
+      end
+    end
+
+    return_element
+  end
+
+  def my_all?
+    # see if all or no values match condition
+    return true unless block_given?
+
+    my_each do |n|
+      # return true unless any of the values does not meet the statement
+      if is_a? Hash
+        return false unless yield n.to_a[0], n.to_a[1]
+      elsif !(yield n)
+        return false
+      end
+    end
+
+    true
+  end
+
+  def my_any?
+    # see if any of the values match condition
+    return true unless block_given?
+
+    my_each do |n|
+      # return true if any of the values meet the statement
+      if is_a? Hash
+        return true if yield n.to_a[0], n.to_a[1]
+      elsif yield n
+        return true
+      end
+    end
+
+    false
+  end
+
+  def my_none?
+    # the reverse of my_all?
+    return false unless block_given?
+
+    my_each do |n|
+      if is_a? Hash
+        return true unless yield n.to_a[0], n.to_a[1]
+      elsif !(yield n)
+        return true
+      end
+    end
+
+    false
+  end
+
+  def my_count
+    return length unless block_given?
+
+    count = 0
+
+    my_each do |n|
+      if is_a? Hash
+        count += 1 if yield n.to_a[0], n.to_a[1]
+      elsif yield n
+        count += 1
+      end
+    end
+
+    count
+  end
+
+  def my_map(proc = nil)
+    return to_enum(:my_map) if !block_given? && proc.nil?
+
+    return_element = []
+
+    if !proc.nil?
+      my_each_with_index do |n, idx|
+        return_element[idx] = proc.call(n)
+      end
+    else
+      my_each_with_index do |n, idx|
+        return_element[idx] = yield n
+      end
+    end
+
+    return_element
+  end
+
+  def my_inject(*parameters_array)
+    return raise 'no block given' unless block_given?
+
+    array = proper_array_for_inject(self)
+
+    if parameters_array.empty?
+      cumulative_variable = array[0]
+      array.shift
+    else
+      cumulative_variable = if (parameters_array[0].is_a? Hash) || (parameters_array[0].is_a? Array)
+                              parameters_array[0].class.new
+                            else
+                              parameters_array[0]
+                            end
+    end
+
+    array.my_each do |n|
+      cumulative_variable = yield cumulative_variable, n
+    end
+
+    cumulative_variable
+  end
```

Comment:
- [ ] should be identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [ ] combines all elements of enum by applying a binary operation, specified by a block
- [ ] when a symbol is specified combines each element of the collection by applying the symbol as a named method